pkg/spartan/plugin.go

```diff
@@ -0,0 +1,170 @@
+package spartan
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types/current"
+)
+
+func setupContainerVeth(netns, ifName string, mtu int, pr current.Result, spartanIPs []net.IPNet) (string, error) {
+	// The IPAM result will be something like IP=192.168.3.5/24,
+	// GW=192.168.3.1. What we want is really a point-to-point link but
+	// veth does not support IFF_POINTOPONT. So we set the veth
+	// interface to 192.168.3.5/32. Since the device netmask is set to
+	// /32, this would not add any routes to the main routing table.
+	// Therefore, in order to reach the spartan interfaces, we will have
+	// to explicitly set routes to the spartan interface through this
+	// device.
+
+	var hostVethName string
+
+	err := ns.WithNetNSPath(netns, func(hostNS ns.NetNS) error {
+		hostVeth, _, err := ip.SetupVeth(ifName, mtu, hostNS)
+		if err != nil {
+			return err
+		}
+
+		containerVeth, err := netlink.LinkByName(ifName)
+		if err != nil {
+			return fmt.Errorf("failed to lookup container VETH %q: %s", ifName, err)
```

Comment:
or if you're wrapping errors (vs `errors.New`) ... you could consider something like https://godoc.org/github.com/pkg/errors (via https://dave.cheney.net/2016/06/12/stack-traces-and-the-errors-package)